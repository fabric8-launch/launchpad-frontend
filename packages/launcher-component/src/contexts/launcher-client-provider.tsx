import React, { useState } from 'react';
import { LauncherClientContext } from './launcher-client-context';
import { AuthorizationTokenProvider, cachedLauncherClient, checkNotNull, LauncherClient, mockLauncherClient } from 'launcher-client';

interface LauncherClientProviderProps {
  children: React.ReactNode;
  client?: LauncherClient;
  creatorUrl?: string;
  launcherUrl?: string;
  authorizationTokenProvider?: AuthorizationTokenProvider;
}

function buildLauncherClient(props: LauncherClientProviderProps) {
  if(props.client) {
    return props.client;
  }
  let client: LauncherClient;
  if (!!props.creatorUrl || !!props.launcherUrl) {
    checkNotNull(props.launcherUrl, 'launcherUrl');
    checkNotNull(props.creatorUrl, 'creatorUrl');
    client = cachedLauncherClient({creatorUrl: props.creatorUrl!, launcherURL: props.launcherUrl!});
  } else {
    client = mockLauncherClient();
  }
  return client;
}

export function LauncherClientProvider(props: LauncherClientProviderProps) {
  const [client] = useState<LauncherClient>(buildLauncherClient(props));

  if (props.authorizationTokenProvider) {
    client.authorizationTokenProvider = props.authorizationTokenProvider;
  }

  return (
    <LauncherClientContext.Provider value={client}>
      {props.children}
    </LauncherClientContext.Provider>
  );
}
