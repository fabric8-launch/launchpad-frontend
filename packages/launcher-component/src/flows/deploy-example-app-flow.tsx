import * as React from 'react';
import { useSessionStorageWithObject } from 'react-use-sessionstorage';

import { createDestRepositoryFormValueWithGeneratedName, DestRepositoryHub } from '../hubs/dest-repository-hub';
import { toExamplePayload } from './launcher-client-adapters';
import { ExampleHub } from '../hubs/example-hub';
import { LaunchFlow, useAutoSetCluster } from './launch-flow';
import { DeploymentHub } from '../hubs/deployment-hub';
import { ExampleApp } from './types';

const defaultExampleApp = {
  example: {},
  destRepository: createDestRepositoryFormValueWithGeneratedName(),
  deployment: {},
};

function getFlowStatus(app: ExampleApp) {
  if (!ExampleHub.checkCompletion(app.example)) {
    return {
      hint: 'You should select an example application.',
      isReadyForDownload: false,
      isReadyForLaunch: false,
    };
  }
  if (!DeploymentHub.checkCompletion(app.deployment)) {
    return {
      hint: 'If you wish to Launch your application, you should configure OpenShift Deployment.',
      isReadyForDownload: true,
      isReadyForLaunch: false,
    };
  }
  if (!DestRepositoryHub.checkCompletion(app.destRepository)) {
    return {
      hint: 'If you wish to Launch your application, you should configure the destination repository.',
      isReadyForDownload: true,
      isReadyForLaunch: false,
    };
  }
  return {
    hint: 'Your application is ready to launch!',
    isReadyForDownload: true,
    isReadyForLaunch: true,
  };
}

export function DeployExampleAppFlow(props: { onCancel?: () => void }) {
  const [app, setApp, clear] = useSessionStorageWithObject<ExampleApp>('app', defaultExampleApp);
  const showDeploymentForm = useAutoSetCluster(setApp);

  const onCancel = () => {
    clear();
    props.onCancel!();
  };

  const flowStatus = getFlowStatus(app);

  const items = [
    {
      id: 'example',
      title: 'Example',
      overview: {
        component: ({edit}) => (
          <ExampleHub.Overview value={app.example} onClick={edit}/>
        ),
        width: 'third',
      },
      form: {
        component: ({close}) => (
          <ExampleHub.Form
            initialValue={app.example}
            onSave={(example) => {
              setApp({...app, example});
              close();
            }}
            onCancel={close}
          />
        ),
      }
    },
    {
      id: 'destRepository',
      title: 'Destination Repository',
      overview: {
        component: ({edit}) => (
          <DestRepositoryHub.Overview value={app.destRepository} onClick={edit}/>
        ),
        width: 'third',
      },
      form: {
        component: ({close}) => (
          <DestRepositoryHub.Form
            initialValue={app.destRepository}
            onSave={(srcLocation) => {
              setApp({...app, destRepository: srcLocation});
              close();
            }}
            onCancel={close}
          />
        ),
      }
    },
    {
      id: 'openshift-deployment',
      title: 'OpenShift Deployment',
      overview: {
        component: ({edit}) => (
          <DeploymentHub.Overview value={app.deployment} onClick={edit}/>
        ),
        width: 'third',
      },
      form: showDeploymentForm && {
        component: ({close}) => (
          <DeploymentHub.Form
            initialValue={app.deployment}
            onSave={(deployment) => {
              setApp({...app, deployment});
              close();
            }}
            onCancel={close}
          />
        ),
      }
    }
  ];

  return (
    <LaunchFlow
      title="Deploy an Example Application"
      items={items}
      {...flowStatus}
      buildAppPayload={() => {
        clear();
        return toExamplePayload(app);
      }}
      onCancel={onCancel}
    />
  );

}
