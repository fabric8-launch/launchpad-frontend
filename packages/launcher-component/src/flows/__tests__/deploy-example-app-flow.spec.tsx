import * as React from 'react';
import 'jest-dom/extend-expect';
import { cleanup, render } from 'react-testing-library';
import { DeployExampleAppFlow } from '../deploy-example-app-flow';
import { flushPromises } from './flow-helpers';
import { LauncherClientProvider } from '../../contexts/launcher-client-provider';

afterEach(() => {
  console.log('cleanup()');
  cleanup();
});

jest.useFakeTimers();

describe('<DeployExampleAppFlow />', () => {
  it('renders and initializes the DeployExampleAppFlow correctly', async () => {
    const comp = render(<LauncherClientProvider><DeployExampleAppFlow appName="my-test-app"/></LauncherClientProvider>);
    expect(comp.getByLabelText('Loading dest-repository')).toBeDefined();
    expect(comp.getByLabelText('Loading openshift-deployment')).toBeDefined();

    // Resolve data from auto loader
    await flushPromises();

    // Resolve overview promises
    await flushPromises();

    expect(comp.getByLabelText('dest-repository is configured')).toBeDefined();
    expect(comp.getByLabelText('openshift-deployment is configured')).toBeDefined();

    expect(comp.getByLabelText('example is not configured')).toBeDefined();

    expect(comp.getByLabelText('Launch Application')).toHaveAttribute('disabled');
    expect(comp.getByLabelText('Download Application')).toHaveAttribute('disabled');
  });
});
