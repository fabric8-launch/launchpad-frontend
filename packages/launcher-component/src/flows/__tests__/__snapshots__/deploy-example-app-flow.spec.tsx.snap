// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeployExampleAppFlow /> Choose example backend and check full launch until next steps popup: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "category": "example",
        "shared": Object {
          "mission": Object {
            "id": "crud",
          },
          "runtime": Object {
            "name": "vert.x",
            "version": "redhat",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;
